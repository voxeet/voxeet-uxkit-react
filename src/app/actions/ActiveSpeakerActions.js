import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const Types = {
  START_ACTIVE_SPEAKER: "START_ACTIVE_SPEAKER",
  STOP_ACTIVE_SPEAKER: "STOP_ACTIVE_SPEAKER",
  PARTICIPANT_SPEAKING: "PARTICIPANT_SPEAKING",
  DISABLE_FORCE_ACTIVE_SPEAKER: "DISABLE_FORCE_ACTIVE_SPEAKER",
  FORCE_ACTIVE_SPEAKER: "FORCE_ACTIVE_SPEAKER",
  SILENCE: "SILENCE",
};

export class Actions {
  static startActiveSpeaker() {
    return (dispatch, getState) => {
      const {
        voxeet: { participants },
      } = getState();
      let interval = participants.interval;
      if (!interval) {
        interval = setInterval(() => {
          const {
            voxeet: { participants, activeSpeaker },
          } = getState();
          if (participants.screenShareEnabled) return;
          if (!activeSpeaker.forceActiveUserEnabled) {
            for (let participant of participants.participants) {
              if (participant.participant_id)
                VoxeetSDK.conference.isSpeaking(
                  VoxeetSDK.conference.participants.get(
                    participant.participant_id
                  ),
                  (status) => {
                    participant.isSpeaking = status;
                  }
                );
            }
            const participantsConnected = participants.participants.filter(
              (p) => p.isConnected
            );
            const activeParticipantConnected = activeSpeaker.activeSpeaker?
                participantsConnected.find(participant =>
                  activeSpeaker.activeSpeaker.participant_id === participant.participant_id
                ):
                false;
            const participant =
              participantsConnected.length === 1
                ? participantsConnected[0]
                : participantsConnected.find((p) => p.isSpeaking) ||
                  null;
            if(participant) {
              // Set new active speaker if there are none
              if(!activeParticipantConnected || !activeSpeaker.activeSpeaker || !activeSpeaker.activeSpeakerSince
                  || activeSpeaker.activeSpeaker.participant_id == participant.participant_id) {
                if(activeSpeaker.activeSpeaker && activeSpeaker.activeSpeaker.participant_id === participant.participant_id) {
                  dispatch({
                    type: Types.PARTICIPANT_SPEAKING,
                    payload: { participant },
                  });
                } else {
                  // console.log('New active speaker', participant);
                  // console.log('!activeParticipantConnected, !activeSpeaker.activeSpeaker, !activeSpeaker.activeSpeakerSince',
                  //    !activeParticipantConnected, !activeSpeaker.activeSpeaker, !activeSpeaker.activeSpeakerSince);
                  dispatch({
                    type: Types.PARTICIPANT_SPEAKING,
                    payload: { participant, activeSpeakerSince: Date.now() },
                  });
                }
              }
              // Set new active speaker
              else if(activeSpeaker.activeSpeaker.participant_id != participant.participant_id &&
                  ( !activeSpeaker.activeSpeakerSince || Date.now()-activeSpeaker.activeSpeakerSince>3000) ) {
                // console.log('Switch to next active speaker', participant);
                dispatch({
                  type: Types.PARTICIPANT_SPEAKING,
                  payload: { participant, activeSpeakerSince: Date.now() },
                });
              } //else {
                // console.log('Delay next active speaker', participant);
              //}
            } else {
              //console.log('No speakers');
              dispatch(this.startSilence())
            }
          }
        }, 500);
      }

      dispatch({
        type: Types.START_ACTIVE_SPEAKER,
        payload: {
          interval,
        },
      });
    };
  }

  static stopActiveSpeaker() {
    return {
      type: Types.STOP_ACTIVE_SPEAKER,
    };
  }

  static startSilence() {
    return {
      type: Types.SILENCE,
    };
  }

  static disableForceActiveSpeaker() {
    return {
      type: Types.DISABLE_FORCE_ACTIVE_SPEAKER,
    };
  }

  static forceActiveSpeaker(participant) {
    return {
      type: Types.FORCE_ACTIVE_SPEAKER,
      payload: { participant },
    };
  }
}
