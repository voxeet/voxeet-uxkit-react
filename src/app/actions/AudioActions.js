import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export class Actions {
    static setCaptureMode(captureMode, noiseReductionLevel) {
        return () => {
            return VoxeetSDK.audio.setCaptureMode({audioMode: captureMode, audioModeOptions: {noiseReductionLevel}});
          };
    }
}