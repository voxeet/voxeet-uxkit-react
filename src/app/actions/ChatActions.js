export const Types = {
  ADD_MESSAGE_CHAT: "ADD_MESSAGE_CHAT",
  RESET_BADGE_MESSAGE_CHAT: "RESET_BADGE_MESSAGE_CHAT",
  CLEAR_MESSAGES: "CLEAR_MESSAGES",
  NEW_BADGE_MESSAGE_CHAT: "NEW_BADGE_MESSAGE_CHAT",
};

export class Actions {
  static addMessage(chat) {
    return {
      type: Types.ADD_MESSAGE_CHAT,
      payload: { chat },
    };
  }

  static clearMessages() {
    return {
      type: Types.CLEAR_MESSAGES,
    };
  }

  static newBadgeMessage() {
    return {
      type: Types.NEW_BADGE_MESSAGE_CHAT,
    };
  }

  static resetBadgeMessage() {
    return {
      type: Types.RESET_BADGE_MESSAGE_CHAT,
    };
  }
}
