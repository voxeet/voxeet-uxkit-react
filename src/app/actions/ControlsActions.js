export const Types = {
  TOGGLE_WIDGET: "TOGGLE_WIDGET",
  TOGGLE_FULLSCREEN: "TOGGLE_FULLSCREEN",
  TOGGLE_MODE: "TOGGLE_MODE",
  TOGGLE_MICROPHONE: "TOGGLE_MICROPHONE",
  TOGGLE_VIDEO: "TOGGLE_VIDEO",
  TOGGLE_AUDIO: "TOGGLE_AUDIO",
  DISPLAY_MODES_ALLOWED: "DISPLAY_MODES_ALLOWED",
  DISPLAY_ACTIONS_ALLOWED: "DISPLAY_ACTIONS_ALLOWED",
  SHARE_ACTIONS_ALLOWED: "SHARE_ACTIONS_ALLOWED",
  TOGGLE_MODAL: "TOGGLE_MODAL",
  FORCE_MODE: "FORCE_MODE",
  TOGGLE_SCREEN_SHARE_MODE: "TOGGLE_SCREEN_SHARE_MODE",
  TOGGLE_FILE_PRESENTATION_MODE: "TOGGLE_FILE_PRESENTATION_MODE",
  TOGGLE_VIDEO_PRESENTATION_MODE: "TOGGLE_VIDEO_PRESENTATION_MODE",
  TOGGLE_AUDIO3D: "TOGGLE_AUDIO3D",
  SET_CHROME_EXTENSION_ID: "SET_CHROME_EXTENSION_ID",
  SET_VIDEO_RATIO: "SET_VIDEO_RATIO",
  TOGGLE_RECORDING: "TOGGLE_RECORDING",
  TOGGLE_MODAL_WIDGET: "TOGGLE_MODAL_WIDGET",
  SAVE_CONSTRAINTS: "SAVE_CONSTRAINTS",
  ADMIN_ACTIVED: "ADMIN_ACTIVED",
  KICK_ON_HANG_UP: "KICK_ON_HANG_UP",
  RESET_WIDGET_CONTROLS: "RESET_WIDGET_CONTROLS",
  TOGGLE_ATTENDEES_LIST: "TOGGLE_ATTENDEES_LIST",
  TOGGLE_ATTENDEES_CHAT: "TOGGLE_ATTENDEES_CHAT",
  TOGGLE_ATTENDEES_SETTINGS: "TOGGLE_ATTENDEES_SETTINGS",
  UNLOCK_RECORDING: "UNLOCK_RECORDING",
  DISABLE_SOUNDS: "DISABLE_SOUNDS",
  LOCK_RECORDING: "LOCK_RECORDING",
  CLOSE_SESSION_HANG_UP: "CLOSE_SESSION_HANG_UP",
  SET_SIMULCAST: "SET_SIMULCAST",
  TOGGLE_AUDIO_TRANSPARENT_MODE: "TOGGLE_AUDIO_TRANSPARENT_MODE",
  SET_AUDIO_TRANSPARENT_MODE: "SET_AUDIO_TRANSPARENT_MODE",
  TOGGLE_MAX_REMOTE_PARTICIPANTS: "TOGGLE_MAX_REMOTE_PARTICIPANTS",
  TOGGLE_REQUESTED_VIDEO: "TOGGLE_REQUESTED_VIDEO",
  SET_REQUESTED_VIDEO: "SET_REQUESTED_VIDEO",
  SET_CONFERENCE_PERMISSIONS: "SET_CONFERENCE_PERMISSIONS",
  SET_VIRTUAL_BACKGROUND_MODE: "SET_VIRTUAL_BACKGROUND_MODE",
};

export class Actions {
  static toggleWidget() {
    return {
      type: Types.TOGGLE_WIDGET
    };
  }

  static setSimulcast(simulcast) {
    return {
      type: Types.SET_SIMULCAST,
      payload: {
        simulcast: simulcast
      }
    }
  }

  static toggleFullScreen() {
    return {
      type: Types.TOGGLE_FULLSCREEN
    };
  }

  static closeSessionAtHangUp(closeSessionAtHangUp) {
    return {
      type: Types.CLOSE_SESSION_HANG_UP,
      payload: {
        closeSessionAtHangUp: closeSessionAtHangUp
      }
    };
  }

  static setChromeExtensionId(chromeExtensionId) {
    return {
      type: Types.SET_CHROME_EXTENSION_ID,
      payload: {
        chromeExtensionId: chromeExtensionId
      }
    };
  }

  static saveConstraints(constraints) {
    return {
      type: Types.SAVE_CONSTRAINTS,
      payload: {
        constraints: constraints
      }
    };
  }

  static disableSounds() {
    return {
      type: Types.DISABLE_SOUNDS
    };
  }

  static setVideoRatio(videoRatio) {
    return {
      type: Types.SET_VIDEO_RATIO,
      payload: {
        videoRatio: videoRatio
      }
    };
  }

  static forceMode(mode) {
    return {
      type: Types.FORCE_MODE,
      payload: {
        mode: mode
      }
    };
  }

  static toggleScreenShareMode(isScreenshare) {
    return {
      type: Types.TOGGLE_SCREEN_SHARE_MODE,
      payload: {
        isScreenshare
      }
    };
  }

  static toggleFilePresentationMode(isFilePresentation) {
    return {
      type: Types.TOGGLE_FILE_PRESENTATION_MODE,
      payload: {
        isFilePresentation
      }
    };
  }

  static toggleVideoPresentationMode(isVideoPresentation) {
    return {
      type: Types.TOGGLE_VIDEO_PRESENTATION_MODE,
      payload: {
        isVideoPresentation
      }
    };
  }

  static toggleMode() {
    return {
      type: Types.TOGGLE_MODE
    };
  }

  static toggleAudio3D() {
    return {
      type: Types.TOGGLE_AUDIO3D
    };
  }

  static adminActived() {
    return {
      type: Types.ADMIN_ACTIVED
    };
  }

  static isKickOnHangUpActived() {
    return {
      type: Types.KICK_ON_HANG_UP
    };
  }

  static displayModesAllowed(displayModes) {
    return {
      type: Types.DISPLAY_MODES_ALLOWED,
      payload: {
        displayModes: displayModes
      }
    };
  }

  static shareActionsAllowed(shareActions) {
    return {
      type: Types.SHARE_ACTIONS_ALLOWED,
      payload: {
        shareActions: shareActions
      }
    };
  }

  static displayActionsAllowed(displayActions) {
    return {
      type: Types.DISPLAY_ACTIONS_ALLOWED,
      payload: {
        displayActions: displayActions
      }
    };
  }

  static toggleAudio(state) {
    return {
      type: Types.TOGGLE_AUDIO,
      payload: {
        state: state
      }
    };
  }

  static toggleMicrophone() {
    return {
      type: Types.TOGGLE_MICROPHONE
    };
  }

  static toggleVideo(state) {
    return {
      type: Types.TOGGLE_VIDEO,
      payload: {
        state: state
      }
    };
  }

  static toggleModal() {
    return {
      type: Types.TOGGLE_MODAL
    };
  }

  static toggleRecording() {
    return {
      type: Types.TOGGLE_RECORDING
    };
  }

  static lockRecording() {
    return {
      type: Types.LOCK_RECORDING
    };
  }

  static unlockRecording() {
    return {
      type: Types.UNLOCK_RECORDING
    };
  }

  static toggleModalWidget() {
    return {
      type: Types.TOGGLE_MODAL_WIDGET
    };
  }

  static resetWidgetControls() {
    return {
      type: Types.RESET_WIDGET_CONTROLS
    };
  }

  static toggleAttendeesList() {
    return {
      type: Types.TOGGLE_ATTENDEES_LIST
    };
  }

  static toggleAttendeesSettings() {
    return {
      type: Types.TOGGLE_ATTENDEES_SETTINGS
    };
  }

  static setAudioTransparentMode(state) {
    return {
      type: Types.SET_AUDIO_TRANSPARENT_MODE,
      payload: {
        audioTransparentMode : state
      }
    };
  }

  static toggleAudioTransparentMode() {
    return {
      type: Types.TOGGLE_AUDIO_TRANSPARENT_MODE
    };
  }

  static setMaxVideoForwarding(state) {
    return {
      type: Types.TOGGLE_MAX_REMOTE_PARTICIPANTS,
      payload: {
        maxVideoForwarding: state
      }
    };
  }

  static toggleAttendeesChat() {
    return {
      type: Types.TOGGLE_ATTENDEES_CHAT
    };
  }

  static toggleRequestedVideo(participant_id) {
    return {
      type: Types.TOGGLE_REQUESTED_VIDEO,
      payload: {
        participant_id
      }
    };
  }

  static setRequestedVideo(participant_id, state) {
    return {
      type: Types.SET_REQUESTED_VIDEO,
      payload: {
        participant_id,
        state
      }
    };
  }

  static setConferencePermissions(conferencePermissions) {
    return {
      type: Types.SET_CONFERENCE_PERMISSIONS,
      payload: {
        conferencePermissions
      }
    };
  }

  static setVirtualBackgroundMode(mode) {
    return {
      type: Types.SET_VIRTUAL_BACKGROUND_MODE,
      payload: {
        mode: mode
      }
    };
  }
}
