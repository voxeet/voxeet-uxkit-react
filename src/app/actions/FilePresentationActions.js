export const Types = {
  FILE_PRESENTATION_CONVERTED_START: "FILE_PRESENTATION_CONVERTED_START",
  FILE_PRESENTATION_ADD_THUMB: "FILE_PRESENTATION_ADD_THUMB",
  FILE_PRESENTATION_CONVERTED_OVER: "FILE_PRESENTATION_CONVERTED_OVER",
  FILE_PRESENTATION_START: "FILE_PRESENTATION_START",
  FILE_PRESENTATION_STOP: "FILE_PRESENTATION_STOP",
  FILE_PRESENTATION_UPDATE: "FILE_PRESENTATION_UPDATE",
};

export class Actions {
  static fileConvertStart() {
    return {
      type: Types.FILE_PRESENTATION_CONVERTED_START,
    };
  }

  static addThumbnail(thumbUrl) {
    return {
      type: Types.FILE_PRESENTATION_ADD_THUMB,
      payload: { thumbUrl },
    };
  }

  static fileConvertStop(filePresentationId) {
    return {
      type: Types.FILE_PRESENTATION_CONVERTED_OVER,
      payload: { filePresentationId },
    };
  }

  static startFilePresentation(
    filePresentationId,
    fileUrl,
    position,
    imageCount
  ) {
    return {
      type: Types.FILE_PRESENTATION_START,
      payload: { filePresentationId, fileUrl, position, imageCount },
    };
  }

  static stopFilePresentation() {
    return {
      type: Types.FILE_PRESENTATION_STOP,
    };
  }

  static updateFilePresentation(position, fileUrl) {
    return {
      type: Types.FILE_PRESENTATION_UPDATE,
      payload: { position, fileUrl },
    };
  }
}
