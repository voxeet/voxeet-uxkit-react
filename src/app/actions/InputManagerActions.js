export const Types = {
  INPUT_AUDIO_CHANGE: "INPUT_AUDIO_CHANGE",
  OUTPUT_AUDIO_CHANGE: "OUTPUT_AUDIO_CHANGE",
  INPUT_VIDEO_CHANGE: "INPUT_VIDEO_CHANGE",
};

export class Actions {
  static inputAudioChange(device) {
    return {
      type: Types.INPUT_AUDIO_CHANGE,
      payload: { device },
    };
  }

  static inputVideoChange(device, isBackCamera) {
    return {
      type: Types.INPUT_VIDEO_CHANGE,
      payload: { device, isBackCamera },
    };
  }

  static outputAudioChange(device) {
    return {
      type: Types.OUTPUT_AUDIO_CHANGE,
      payload: { device },
    };
  }
}
