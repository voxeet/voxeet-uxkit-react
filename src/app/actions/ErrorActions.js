import { Actions as ControlsActions } from "./ControlsActions";

export const Types = {
  ERROR: "ERROR",
  CLEAR_ERROR: "CLEAR_ERROR",
};

export class Actions {
  static onError(error) {
    return {
      type: Types.ERROR,
      payload: { error },
    };
  }

  static onClearError() {
    return {
      type: Types.CLEAR_ERROR,
    };
  }
}
