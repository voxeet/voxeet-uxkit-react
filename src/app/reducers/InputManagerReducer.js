import { Types } from "../actions/InputManagerActions";

const defaultState = {
  currentAudioDevice: "",
  currentOutputDevice: "",
  currentVideoDevice: "",
  isBackCamera: false,
};

const InputManagerReducer = (state = defaultState, action) => {
  switch (action.type) {
    case Types.INPUT_AUDIO_CHANGE:
      return {
        ...state,
        currentAudioDevice: action.payload.device
      };
    case Types.OUTPUT_AUDIO_CHANGE:
      return {
        ...state,
        currentOutputDevice: action.payload.device
      };
    case Types.INPUT_VIDEO_CHANGE:
      return {
        ...state,
        isBackCamera: action.payload.isBackCamera,
        currentVideoDevice: action.payload.device
      };
    default:
      return state;
  }
};

export default InputManagerReducer;