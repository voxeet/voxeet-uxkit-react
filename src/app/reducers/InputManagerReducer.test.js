import InputManagerReducer from "./InputManagerReducer";
import { Types } from "../actions/InputManagerActions";

describe("input manager reducer test suite", () => {
  test("reducer should return default state", () => {
    //1- arrange
    const action = { type: "" };

    //2- act
    const state = InputManagerReducer({}, action);

    //3- assert
    expect(state).toMatchSnapshot();
  });

  test("state should match snapshot when handling audio input update", () => {
    const device = {
      deviceId: "123",
      label: "audio input",
      kind: "audioinput",
      groupId: "123",
    };
    const action = {
      type: Types.INPUT_AUDIO_CHANGE,
      payload: {
        device,
      },
    };

    const state = InputManagerReducer({}, action);

    expect(state).toMatchSnapshot();
  });

  test("state should match snapshot when handling audio output update", () => {
    const device = {
      deviceId: "123",
      label: "audio output",
      kind: "audiooutput",
      groupId: "123",
    };
    const action = {
      type: Types.OUTPUT_AUDIO_CHANGE,
      payload: {
        device,
      },
    };

    const state = InputManagerReducer({}, action);

    expect(state).toMatchSnapshot();
  });

  test("state should match snapshot when handling video input update", () => {
    const device = {
      deviceId: "123",
      label: "video input",
      kind: "videoinput",
      groupId: "123",
    };
    const action = {
      type: Types.INPUT_VIDEO_CHANGE,
      payload: {
        device,
      },
    };

    const state = InputManagerReducer({}, action);

    expect(state).toMatchSnapshot();
  });
});
