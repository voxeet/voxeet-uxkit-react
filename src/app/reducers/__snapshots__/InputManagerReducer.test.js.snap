// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input manager reducer test suite reducer should return default state 1`] = `
Object {
  "currentAudioDevice": "",
  "currentOutputDevice": "",
  "currentVideoDevice": "",
  "isBackCamera": false,
}
`;

exports[`input manager reducer test suite state should match snapshot when handling audio input update 1`] = `
Object {
  "currentAudioDevice": Object {
    "deviceId": "123",
    "groupId": "123",
    "kind": "audioinput",
    "label": "audio input",
  },
}
`;

exports[`input manager reducer test suite state should match snapshot when handling audio output update 1`] = `
Object {
  "currentOutputDevice": Object {
    "deviceId": "123",
    "groupId": "123",
    "kind": "audiooutput",
    "label": "audio output",
  },
}
`;

exports[`input manager reducer test suite state should match snapshot when handling video input update 1`] = `
Object {
  "currentVideoDevice": Object {
    "deviceId": "123",
    "groupId": "123",
    "kind": "videoinput",
    "label": "video input",
  },
  "isBackCamera": undefined,
}
`;
